### ISSUE TITLE
Idea: promise polyfill based flow

### ISSUE THREAD

**Issue by xaviergonz** [mobxjs/mobx#1726]
Welcome to MobX. Please provide as much relevant information as possible!

I have a:

3. [X] Idea:
  * [X] What problem would it solve for you?
  * [X] Do you think others will benefit from this change as well and it should in core package (see also mobx-utils)?
  * [X] Are you willing to (attempt) a PR yourself?

Given the problems with typescript with yield (it always returns any, having to typecast the return type) I started researching it it would be possible to have a "flow" behaviour with plain async/await, and I think I found a solution:

Given the following proof of concept
```ts
// library code

let runningInAction = 0;

function runInAction<T>(f: () => T): T {
  try {
    runningInAction++;
    console.log("ria start", runningInAction);
    return f();
  } finally {
    console.log("ria end", runningInAction);
    runningInAction--;
  }
}

function tweakPromise() {
  const oldThen = Promise.prototype.then;
  const oldCatch = Promise.prototype.catch;

  Promise.prototype.then = function(onfulfill?, onreject?) {
    const self = this;
    let wrappedOnfulfill = onfulfill;
    let wrappedOnreject = onreject;

    if (runningInAction > 0) {
      if (onfulfill) {
        wrappedOnfulfill = function(...args) {
          return runInAction(() => {
            return onfulfill.apply(self, args);
          });
        };
      }

      if (onreject) {
        wrappedOnreject = function(...args) {
          return runInAction(() => {
            return onreject.apply(self, args);
          });
        };
      }
    }

    return oldThen.call(self, wrappedOnfulfill, wrappedOnreject);
  };

  Promise.prototype.catch = function(onreject?) {
    const self = this;
    let wrappedOnreject = onreject;

    if (runningInAction) {
      if (onreject) {
        wrappedOnreject = function(...args) {
          return runInAction(() => {
            return onreject.apply(self, args);
          });
        };
      }
    }

    return oldCatch.call(self, wrappedOnreject);
  };
}

tweakPromise();

// user code

async function b() {
  console.log("b.0");
  await new Promise(r => setTimeout(r, 1000));
  console.log("b.1");
  return 1;
}

async function a() {
  return await runInAction(async () => {
    console.log("a.0");
    const res = (await b()) + 2;
    console.log("a.1");
    return res;
  });
}

async function outsideRia() {
  await new Promise(r => {
    setTimeout(() => {
      console.log("I should be outside ria");
      r();
    }),
      100;
  });
}

async function main() {
  console.log("should be run without run in action stuff");
  await outsideRia();
  console.log("---------");

  console.log("should be run with run in action stuff");
  const pa1 = a();
  const pa2 = a();

  // promises created outside the run in action, bit in the middle of the process should not have ria wrappers
  await outsideRia();

  const res1 = await pa1;
  const res2 = await pa2;
  console.log("---------");
  console.log("res1", res1);
  console.log("res2", res2);

  console.log("should be run without run in action stuff");
  await outsideRia();
  console.log("---------");
}

main().then(() => {
  console.log("done");
});
```

The following output (plus my own comments) is generated:
```
should be run without run in action stuff
// as you see, this promise is outside run in action since it is was run outside an action
I should be outside ria 
---------

should be run with run in action stuff
// the start of the actions are run inside a run in action context
ria start 1
a.0
b.0
ria end 1
ria start 1
a.0
b.0
ria end 1
// a promise generated in the middle of an action (but while the action is inactive) is properly run outside a run in action context
I should be outside ria
// the continuations of the actions are properly run once more inside a run in action context
ria start 1
b.1
ria end 1
ria start 1
b.1
ria end 1
ria start 1
a.1
ria end 1
ria start 1
a.1
ria end 1
---------
// once more we are outside a run in action context here
res1 3
res2 3
should be run without run in action stuff
// next promises are also run outside a run in action context
I should be outside ria
---------
done
```

Basically this idea would allow people to write (both in JS and TS) stuff like
```ts
@observable x = 5
@action
async someAsyncAction() {
  // all of this below (even the code of the promise itself) is run inside an action!
  this.observable = 4
  this.observable = await gimmeANumber();
  this.observable = 6
}
```
without loosing strict mode and with proper typings that currently flow doesn't allow

The only drawback is the small modification to the prototype of the promise, but I don't think it is a big deal (and could be opt-in if invoked in configure with some parameter)

Thoughts?

**Comment by xaviergonz** [comment]
Ok, (big) drawback though, using the native async/await of node seems not to work (it seems almost like the "then" part of the promise is executed after the function rather than in the function)

**Comment by urugator** [comment]
https://stackoverflow.com/a/49120177

**Comment by mweststrate** [comment]
@xaviergonz investigated this part earlier, but trying to patch promises is really opening pandora's box. If you really want, @benjamingr can tell you horror stories that will keep you awake at night  :-P

**Comment by benjamingr** [comment]
What Michel said, we can only do this with asyncHooks which are not available in browsers.

I can gladly bring any suggestions or hooks you want to TC39 though
